table/transaction.go

```diff
@@ -142,6 +143,120 @@ func (t *Transaction) SetProperties(props iceberg.Properties) error {
 	return nil
 }
 
+type expireSnapshotsCfg struct {
+	minSnapshotsToKeep *int
+	maxSnapshotAgeMs   *int64
+}
+
+type ExpireSnapshotsOpt func(*expireSnapshotsCfg)
+
+func WithRetainLast(n int) ExpireSnapshotsOpt {
+	return func(cfg *expireSnapshotsCfg) {
+		cfg.minSnapshotsToKeep = &n
+	}
+}
+
+func WithOlderThan(t time.Duration) ExpireSnapshotsOpt {
+	return func(cfg *expireSnapshotsCfg) {
+		n := t.Milliseconds()
+		cfg.maxSnapshotAgeMs = &n
+	}
+}
+
+func (t *Transaction) ExpireSnapshots(opts ...ExpireSnapshotsOpt) error {
+	var (
+		cfg          expireSnapshotsCfg
+		updates      []Update
+		snapsToKeep  = make(map[int64]struct{})
+		refsToDelete = make(map[string]struct{})
+	)
+
+	for _, opt := range opts {
+		opt(&cfg)
+	}
+
+	for refName, ref := range t.meta.refs {
+		if refName == MainBranch {
+			continue
+		}
```

Comment:
we can't remove snapshots from the main branch?

Reply:
As of:  https://iceberg.apache.org/spec/#snapshot-retention-policy

> When expiring snapshots, retention policies in table and snapshot references are evaluated in the following way:

Start with an empty set of snapshots to retain
Remove any refs (other than main) where the referenced snapshot is older than max-ref-age-ms
For each branch and tag, add the referenced snapshot to the retained set
For each branch, add its ancestors to the retained set until:
The snapshot is older than max-snapshot-age-ms, AND
The snapshot is not one of the first min-snapshots-to-keep in the branch (including the branch's referenced snapshot)
Expire any snapshot not in the set of snapshots to retain.